app/src/main/java/com/islamversity/reyan/service/MyFirebaseMessagingService.kt

```diff
@@ -0,0 +1,101 @@
+package com.islamversity.reyan.service
+
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.media.RingtoneManager
+import android.os.Build
+import android.os.Bundle
+import androidx.core.app.NotificationCompat
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import com.islamversity.reyan.MainActivity
+import com.islamversity.reyan.R
+import com.islamversity.reyan.service.NotificationDataType.Companion.NOTIFICATION_DATA_KEY
+import timber.log.Timber
+
+
+class MyFirebaseMessagingService: FirebaseMessagingService() {
+
+
+    override fun onNewToken(token: String) {
+        Timber.d("Notification : Refreshed token: $token")
+        sendFCMTokenToServer(token)
+    }
+
+    private fun sendFCMTokenToServer(token: String) {
+        // Cache.set(CacheKeys.Firebase.FCM_TOKEN_FIREBASE, token)
+        Timber.d("Notification : sendRegistrationTokenToServer($token)")
+    }
+
+    override fun onMessageReceived(remoteMessage: RemoteMessage) {
+        Timber.d("Notification : From: ${remoteMessage.from}")
+
+        val data  = remoteMessage.data
+
+        // Check if message contains a data payload.
+        if (data.isNotEmpty()) {
+            Timber.d("Notification : Message data payload: $data")
+
+            val bundle = Bundle()
+            for ((key, value) in data.entries) {
+                bundle.putString(key, value)
+            }
+
+            showNotification(
+                data["title"],
+                data["body"],
+                bundle
+            )
+        }
+
+        // Check if message contains a notification payload.
+        remoteMessage.notification?.let { notification ->
+            Timber.d("Notification : 2-Message Notification Body: ${notification.body}")
+        }
+    }
+
+    private fun showNotification(title: String?, body: String?, dataBundle: Bundle) {
+
+        val intent = Intent(this, MainActivity::class.java).apply {
+            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
```

Comment:
I think we should add `SingleInstance` as well

Reply:
is it a FLAG ?

Reply:
yeah
asking because if the app is already runing 
with these two we might actually launch a second Task (the first one will be head less)

Reply:
How to do it? I do not find the flag.

Reply:
these  two flags  are used based on document. FLAG_ACTIVITY_NEW_TASK is for SingleInstance on runtime.